scripts/dats_jsonld_annotator/annotator.py

```diff
@@ -0,0 +1,323 @@
+                v, _schema, specific_context, context_dir
+            )
+            specific_context.update(_local_context)
+        if isinstance(v, list):
+            # Let's find the schema link under "items"
+            annotation_list = []
+            for vv in v:
+                if not isinstance(vv, dict):
+                    annotation_list.append(vv)
+                    continue
+                _schema = find_schema(schema, k, vv)
+                _json_o, _local_context = annotate_dats_object(
+                    vv, _schema, specific_context, context_dir
+                )
+                annotation_list.append(_json_o)
+                specific_context.update(_local_context)
+            json_object[k] = annotation_list
+        # Add the key mapping to the context
+        if k not in context.keys():
+            logger.debug(f'{k} not in context: {pal.Path(schema["id"]).name}')
+            continue
+        if k not in specific_context.keys():
+            specific_context[k] = context[k]
+            logger.debug(f"{k} added to context: {context[k]}")
+        elif specific_context[k] == context[k]:
+            continue
+        else:
+            logger.debug(f"{k} duplicate: {specific_context[k]} vs {context[k]}")
+            continue
+
+    dtype = schema["properties"]["@type"]["enum"][0]
+    if dtype not in context.keys():
+        logger.debug(f'{dtype} not in context {schema["id"]}')
+    else:
+        specific_context[dtype] = context[dtype]
+    json_object["@type"] = dtype
+    return json_object, specific_context
+
+
+def gen_jsonld_outpath(dats_json_f, out_path):
+    """
+    This function generates a output file path for the annotated DATS jsonld file.
+
+    :param dats_json_f: the path to the original DATS instance file
+    :param out_path: the folder or new file path where the annotated DATS.jsonld file should be stored
+    :return:  the final DATS.jsonld file path
+    """
+    if not isinstance(dats_json_f, pal.Path):
+        dats_json_f = pal.Path(dats_json_f)
+    if out_path is not None and not isinstance(out_path, pal.Path):
+        out_path = pal.Path(out_path)
+
+    out_name = dats_json_f.with_suffix(".jsonld").name
+    if out_path is None:
+        # We save the output to the parent directory of the current DATS JSON file
+        out_path = dats_json_f.parent
+    elif out_path.suffix != "":
+        # We have most likely gotten a path to a non-existent file, let's use that
+        return out_path
+    elif not out_path == dats_json_f.parent:
+        # We are saving this somewhere other than the original folder
+        out_name = f"{dats_json_f.parent.name}_{out_name}"
+    if out_path.is_dir():
+        dats_jsonld_f = out_path / out_name
+    elif out_path.parent.is_dir():
+        # TODO: check if there is a way to explicitly look for a file that doesn't exist
+        logger.info(f"I will write to {out_path.resolve() = }")
+        dats_jsonld_f = out_path.parent / out_name
```

Comment:
Does this work? I usually use the following (which avoids // in the path):
```suggestion
        dats_jsonld_f = os.path.join(out_path.parent, out_name)
```

Reply:
Yes, both would work. I am using [pathlib](https://docs.python.org/3/library/pathlib.html) in this script because I find it makes for more intuitive syntax.